### PR TITLE
Support case expression

### PR DESCRIPTION
**Pull request by jievince**
http://jira.nebula-graph.io:8081/projects/NG/issues/NG-281?filter=allissues
Based on:
https://github.com/vesoft-inc/nebula-common/pull/269
https://github.com/vesoft-inc/nebula-storage/pull/183

### CODE REVIEW

File: src/parser/parser.yy [hunk 1]
```diff
@@ -557,6 +568,65 @@ attribute_expression
...
     ;
 
+case_expression
+    : simple_case_expression {
+        $$ = $1;
+    }
+    | generic_case_expression {
+        $$ = $1;
+    }
+    | conditional_expression {
```
Comment (Shylock-Hg) [inline src/parser/parser.yy:578]: Is this part of case_expression?

Comment (jievince) [reply src/parser/parser.yy:586]: Yes

Comment (Shylock-Hg) [reply src/parser/parser.yy:589]: But I don't see this in the document of case expression.

Comment (jievince) [reply src/parser/parser.yy:589]: It's here: 
image
conditional_expression is newly added in nebula. Cypher doesn't hava it.

File: tests/query/v2/test_case_expression.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+
+from tests.common.nebula_test_suite import NebulaTestSuite
+from tests.common.nebula_test_suite import T_NULL, T_EMPTY
+import pytest
+
+
+class TestGoQuery(NebulaTestSuite):
```
Comment (Shylock-Hg) [inline tests/query/v2/test_case_expression.py:13]: Change the class name

Comment (jievince) [reply tests/query/v2/test_case_expression.py:13]: Thanks

**Review by CPWstatic — DISMISSED** [review]

**Comment by CPWstatic** [comment]
@jievince @randomJoe211 Docs required.

**Review by nevermore3 — APPROVED** [review]

File: src/visitor/DeduceTypeVisitor.cpp [hunk 3]
```diff
@@ -377,6 +377,32 @@ void DeduceTypeVisitor::visit(EdgeExpression *) {
     type_ = Value::Type::EDGE;
 }
 
+void DeduceTypeVisitor::visit(CaseExpression *expr) {
+    if (expr->hasCondition()) {
+        expr->condition()->accept(this);
```
Comment (yixinglu) [inline src/visitor/DeduceTypeVisitor.cpp:382]: need to fold condition and default expr

**Review by nevermore3 — APPROVED** [review]

File: src/visitor/FoldConstantExprVisitor.cpp [hunk 4]
```diff
@@ -159,6 +159,47 @@ void FoldConstantExprVisitor::visit(MapExpression *expr) {
...
 
+// case Expression
+void FoldConstantExprVisitor::visit(CaseExpression *expr) {
+    bool canBeFolded = true;
+    if (expr->hasCondition()) {
+        expr->condition()->accept(this);
+        if (!canBeFolded_) {
+            canBeFolded = false;
+        }
+        expr->setCondition(fold(expr->condition()));
```
Comment (yixinglu) [inline src/visitor/FoldConstantExprVisitor.cpp:170]: put it in else branch

File: src/visitor/FoldConstantExprVisitor.cpp [hunk 5]
```diff
@@ -159,6 +159,47 @@ void FoldConstantExprVisitor::visit(MapExpression *expr) {
...
+    }
+    if (expr->hasDefault()) {
+        expr->defaultResult()->accept(this);
+        if (!canBeFolded_) {
+            canBeFolded = false;
+        }
+        expr->setDefault(fold(expr->defaultResult()));
```
Comment (yixinglu) [inline src/visitor/FoldConstantExprVisitor.cpp:177]: ditto

**Review by yixinglu — COMMENTED** [review]
Need more test cases:

CASE when 
    case when
     else
    end
else
end

case when end ? case when end : case when end

case when a ? b: c end

**Comment by jievince** [comment]
CI failed due to https://github.com/vesoft-inc/nebula-graph/pull/316

**Review by CPWstatic — APPROVED** [review]

**Review by nevermore3 — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]
excellent

**Comment by yixinglu** [comment]
@randomJoe211 need documents about  case expression

**Pull request merged**